Encryption.md

```diff
@@ -132,49 +203,164 @@ the encryption buffer is comprised of an IV, ciphertext and tag, as described in
+caller code, and passed explicitly to Parquet for the footer and each encrypted column.
+
+### Protection of column metadata
+
+The Parquet file footer, and its nested structures, contain sensitive information - ranging 
 from a secret data (column statistics) to other information that can be exploited by an 
 attacker (e.g. schema, num_values, key_value_metadata, column data offset and size, encoding and crypto_meta_data). 
 This information is automatically protected when the footer and secret columns are encrypted 
 with the same key. In other cases - when column(s) and the footer are encrypted with 
 different keys; or column(s) are encrypted and the footer is not - an extra measure is 
 required to protect the column-specific information in the file footer. In these cases, 
 the column-specific information (kept in `ColumnMetaData` structures) is split from the 
-footer, by utilizing the `required i64 file_offset` parameter in the `ColumnChunk` 
-structure. This allows to serialize each `ColumnMetaData` structure separately, and encrypt 
+footer by utilizing the `required i64 file_offset` parameter in the `ColumnChunk` 
+structure. This allows to serialize each `ColumnMetaData` structure separately and encrypt 
 it with a column-specific key, thus protecting the column stats and other metadata. 
 
+```c
+struct ColumnChunk {
+...
+  /** Byte offset in file_path to the ColumnMetaData **/
+  2: required i64 file_offset
+
+  /** Column metadata for this chunk. This is the same content as what is at
+   * file_path/file_offset.  Having it here has it replicated in the file
+   * metadata.
+   **/
+  3: optional ColumnMetaData meta_data
+..
+}
+```
+
 ### Encrypted footer mode
 
 In files with sensitive column data, a good security practice is to encrypt not only the 
-secret columns, but also the file footer metadata, with a separate footer key. This hides
+secret columns, but also the file footer metadata with a separate footer key. This hides
 the file schema / column names, number of rows, key-value properties, column sort order, 
 column data offset and size, list of encrypted columns and metadata of the column encryption keys. 
-It also makes the footer tamper-proof.
+It also makes the footer tamper-proof and helps to prevent module replacement attacks.
 
-The columns encrypted with the same key as the footer, don't split the ColumnMetaData from the 
+The columns encrypted with the same key as the footer don't separate the ColumnMetaData from the 
 ColumnChunks, leaving it at the original location, `optional ColumnMetaData meta_data`. This field
-is not set for columns enrypted with a column-specific key.
+is not set for columns encrypted with a column-specific key.
 
 A Thrift-serialized `FileCryptoMetaData` structure is written after the footer. It contains 
 information on the file encryption algorithm and on the footer (offset in 
 the file, and an optional key metadata, with a maximal length of 512). Then 
```

Comment:
Several places note that the maximum length of the key metadata is 512 somethings. This should not be duplicated. Key metadata should be documented in one place and referenced.